File: integrations/php/oicana-php/src/ExportFormat.php
<?php

declare(strict_types=1);

namespace Oicana;

/**
 * Export format configuration for template compilation.
 *
 * This class provides type-safe factory methods for creating export format configurations.
 *
 * Example:
 * ```php
 * // PDF export
 * $pdf = $template->compile(exportFormat: ExportFormat::pdf());
 *
 * // PNG export with custom resolution
 * $png = $template->compile(exportFormat: ExportFormat::png(pixelsPerPt: 3.0));
 *
 * // SVG export
 * $svg = $template->compile(exportFormat: ExportFormat::svg());
 * ```
 */
final readonly class ExportFormat
{
    /**
     * @param array{format: string, pixelsPerPt?: float} $config
     */
    private function __construct(
        private array $config
    ) {
    }

    /**
     * Create PDF export format configuration.
     *
     * PDF is the default and most common export format.
     */
    public static function pdf(): self
    {
        return new self(['format' => 'pdf']);
    }

    /**
     * Create SVG export format configuration.
     *
     * SVG is useful for web embedding and vector graphics.
     */
    public static function svg(): self
    {
        return new self(['format' => 'svg']);
    }

    /**
     * Create PNG export format configuration.
     *
     * @param float $pixelsPerPt Resolution multiplier. Higher values produce larger images.
     *                           - 1.0 = 72 DPI (standard)
     *                           - 2.0 = 144 DPI (retina)
     *                           - 3.0 = 216 DPI (high quality print)
     */
    public static function png(float $pixelsPerPt = 2.0): self
    {
        return new self([
            'format' => 'png',
            'pixelsPerPt' => $pixelsPerPt
        ]);
    }

    /**
     * Get the configuration array for the native extension.
     *
     * @return array{format: string, pixelsPerPt?: float}
     */
    public function toArray(): array
    {
        return $this->config;
    }
}

File: integrations/csharp/Oicana/Config/ExportFormat.cs
namespace Oicana.Config;

/// <summary>
/// Export format configuration for an Oicana template compilation.
/// </summary>
public class ExportFormat
{
    internal ExportTarget ExportTarget;
    internal float? PixelsPerPt;

    /// <summary>
    /// Create a new export format configuration for compiling to PDF
    /// </summary>
    public static ExportFormat Pdf()
    {
        return new ExportFormat()
        {
            ExportTarget = ExportTarget.Pdf,
        };
    }

    /// <summary>
    /// Create a new export format configuration for compiling to PNG
    /// </summary>
    /// <param name="pixelsPerPt">The number of pixels per pt. Higher numbers take longer, but create sharper images.</param>
    /// <exception cref="ArgumentOutOfRangeException">If <paramref name="pixelsPerPt"/> is not a positive, finite number.</exception>
    public static ExportFormat Png(float pixelsPerPt = 1.0f)
    {
        if (!float.IsFinite(pixelsPerPt) || pixelsPerPt <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(pixelsPerPt), pixelsPerPt,
                "pixelsPerPt must be a positive, finite number.");
        }

        return new ExportFormat()
        {
            ExportTarget = ExportTarget.Png,
            PixelsPerPt = pixelsPerPt
        };
    }

    /// <summary>
    /// Create a new export format configuration for compiling to SVG
    /// </summary>
    public static ExportFormat Svg()
    {
        return new ExportFormat()
        {
            ExportTarget = ExportTarget.Svg,
        };
    }
}
